test/rate_limiter_test.cc

```diff
@@ -191,43 +191,80 @@ TEST_F(RateLimiterTest, DistributionSamplingRateLimiterImplTest) {
 }
 
 // A rate limiter determines when acquisition is allowed, but DistributionSamplingRateLimiterImpl
-// may arbitrarily delay that. We test that principle here.
-TEST_F(RateLimiterTest, DistributionSamplingRateLimiterImplSchedulingTest) {
-  auto mock_rate_limiter = std::make_unique<NiceMock<MockRateLimiter>>();
-  MockRateLimiter& unsafe_mock_rate_limiter = *mock_rate_limiter;
-  Envoy::Event::SimulatedTimeSystem time_system;
-  auto* unsafe_discrete_numeric_distribution_sampler = new MockDiscreteNumericDistributionSampler();
-  RateLimiterPtr rate_limiter = std::make_unique<DistributionSamplingRateLimiterImpl>(
-      std::unique_ptr<DiscreteNumericDistributionSampler>(
-          unsafe_discrete_numeric_distribution_sampler),
-      std::move(mock_rate_limiter));
-  EXPECT_CALL(unsafe_mock_rate_limiter, timeSource)
-      .Times(AtLeast(1))
-      .WillRepeatedly(ReturnRef(time_system));
+// may arbitrarily delay that. We test that principle with tests that use this fixture, which
+// sets up a distribution sampling rate limiter instance to encapsulate a mock rate limiter,
+// relying on simulated time and a mock discrete numberic distribution sampler.
+class DistributionSamplingRateLimiterTest : public RateLimiterTest {
+public:
+  DistributionSamplingRateLimiterTest()
+      : mock_rate_limiter_(std::make_unique<NiceMock<MockRateLimiter>>()),
+        unsafe_mock_rate_limiter_(*mock_rate_limiter_),
+        unsafe_discrete_numeric_distribution_sampler_(new MockDiscreteNumericDistributionSampler()),
```

Comment:
Is there a special need for bare `new`, or would making the field `std::unique_ptr<MockDiscreteNumericDistributionSampler> discrete_numeric_distribution_sampler_` and referring to `*discrete_numeric_distribution_sampler_` in the tests work just the same? This seems like it like it would get flagged by asan.

Reply:
This ends up being owned by the rate limiter, so there shouldn't be an ASAN report about it. 
Annoyingly, we need temporary member variables to grab references to the mock before we transfer ownership, as the constructor requires a unique_ptr&&. This works nice in prod code by making the path of least resistance a good one, but it is annoying here as we need to go through this hoop. I renamed some things, and made this code use std::move like we do with the rate limiter mocks, to hopefully clean this up and make it more easy to grok.